electron/syntheticsGenerator.ts

```diff
@@ -0,0 +1,364 @@
+    }
+    case 'fill':
+      return `Fill ${action.selector}`;
+    case 'setInputFiles':
+      if (action.files?.length === 0) return `Clear selected files`;
+      else return `Upload ${action.files?.join(', ')}`;
+    case 'navigate':
+      return `Go to ${action.url}`;
+    case 'press':
+      return `Press ${action.key}` + (action.modifiers ? ' with modifiers' : '');
+    case 'select':
+      return `Select ${action.options?.join(', ')}`;
+    case 'assert':
+      return `Assert ${action.selector} ${action.command}`;
+  }
+}
+
+export class SyntheticsGenerator extends PlaywrightGenerator.JavaScriptLanguageGenerator {
+  private previousContext?: ActionInContext;
+  private insideStep: boolean;
+  private varsToHoist: string[];
+
+  constructor(private isSuite: boolean) {
+    super(true);
+    this.insideStep = false;
+    this.previousContext = undefined;
+    this.varsToHoist = [];
+  }
+
+  /**
+   * Generate code for an action.
+   * @param actionInContext The action to create code for.
+   * @returns the strings generated for the action.
+   */
+  generateAction(actionInContext: ActionInContext) {
+    const { action, pageAlias } = actionInContext;
+    if (action.name === 'openPage') {
+      return '';
+    }
+    // Don't cleanup page object managed by Synthetics
+    const isCleanUp = action.name === 'closePage' && pageAlias === 'page';
+    if (isCleanUp) {
+      return '';
+    }
+
+    const stepIndent = this.insideStep ? 2 : 0;
+    const offset = this.isSuite ? 2 + stepIndent : 0 + stepIndent;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(offset);
+
+    const subject = actionInContext.isMainFrame
+      ? pageAlias
+      : actionInContext.frameName
+      ? `${pageAlias}.frame(${formatObject({
+          name: actionInContext.frameName,
+        })})`
+      : `${pageAlias}.frame(${formatObject({
+          url: actionInContext.frameUrl,
+        })})`;
+
+    const signals = toSignalMap(action);
+
+    const isVarHoisted = signals.popup?.popupAlias && this.isVarHoisted(signals.popup.popupAlias);
+    if (signals.dialog) {
+      formatter.add(`  ${pageAlias}.once('dialog', dialog => {
+    console.log(\`Dialog message: $\{dialog.message()}\`);
+    dialog.dismiss().catch(() => {});
+  });`);
+    }
+
+    const emitPromiseAll = signals.waitForNavigation || signals.popup || signals.download;
+    if (emitPromiseAll) {
+      // Generate either await Promise.all([]) or
+      // const [popup1] = await Promise.all([]).
+      let leftHandSide = '';
+      if (signals.popup && signals.popup.popupAlias && !isVarHoisted) {
+        leftHandSide = `const [${signals.popup.popupAlias}] = `;
+      } else if (isVarHoisted) {
+        leftHandSide = `${signals.popup?.popupAlias} = `;
+      } else if (signals.download) {
+        leftHandSide = `const [download] = `;
+      }
+      const rightHandSide = isVarHoisted ? `(await Promise.all([` : `await Promise.all([`;
```

Comment:
Any reason why we are doing this instead of array assignment? 
```ts
[page] = await Promise.all([]) 

// vs

page = (await Promise.all([]))[0]
```

I would keep out formatter close to the PW's to keep the changes minimal and move if we support the pw-test-runner in the future. Also I feel the former is cleaner.

Reply:
https://github.com/elastic/synthetics-recorder/pull/211/commits/52ab47b985240618ef9cee45f2bc855cca750ee7